File: pset1/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

bool luhns_algorathim(char str[]);
void credit_card(char str[]);
bool count_number(long number);
int get_length(long number);

int main(void)
{
    long number;
    do
    {
        number = get_long("Enter your credit card number (only digits): ");
    }
    while ((number < 0));

    char str[25];
    sprintf(str, "%ld", number);

    bool iscount;
    iscount = count_number(number);
    if (iscount)
    {
        bool is_luhns_algorathim = luhns_algorathim(str);
        if (is_luhns_algorathim)
        {
            credit_card(str);
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

bool luhns_algorathim(char str[])
{
    int sum = 0;
    int len = strlen(str);

    for (int i = 0; i < len; i++)
    {
        int digit = str[i] - '0';
        if ((len - i) % 2 == 0)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
    }
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

void credit_card(char str[])
{
    int len = strlen(str);

    if ((len == 16 || len == 13) && (str[0] == '4'))
    {
        printf("VISA\n");
    }
    else if (len == 16 && str[0] == '5' && (str[1] >= '1' && str[1] <= '5'))
    {
        printf("MASTERCARD\n");
    }
    else if (len == 15 && str[0] == '3' && (str[1] == '4' || str[1] == '7'))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

bool count_number(long number)
{
    int len = get_length(number);
    if (len == 13 || len == 15 || len == 16)
    {
        return true;
    }
    else
    {
        return false;
    }
}

int get_length(long number)
{
    int length = 0;
    while (number != 0)
    {
        number /= 10;
        length++;
    }
    return length;
}

File: pset6/credit.py
from cs50 import get_int


# Function for implemeting Luhn's Algorithm
def luhns_algorithm(card_str):
    total_sum = 0
    length = len(card_str)

    for i in range(length):
        digit = int(card_str[i])

        # Check if this digit is in an even position (from the right)
        if (length - i) % 2 == 0:
            digit *= 2
            if digit > 9:
                digit -= 9
        total_sum += digit

    return total_sum % 10 == 0


# Function to find card type
def credit_card_type(card_str):
    length = len(card_str)

    # VISA: 13 or 16 digits and starts with 4
    if length in (13, 16) and card_str.startswith("4"):
        print("VISA")

    # MASTERCARD: 16 digits and starts with 51-55
    elif length == 16 and card_str[:2] in ("51", "52", "53", "54", "55"):
        print("MASTERCARD")

    # AMEX: 15 digits and starts with 34 or 37
    elif length == 15 and card_str[:2] in ("34", "37"):
        print("AMEX")

    else:
        print("INVALID")

# Function for finding length


def valid_length(number):
    length = len(str(number))
    return length in [13, 15, 16]


# Prompt user for valid input
while True:
    number = get_int("Enter your credit card number (only digits): ")
    if number >= 0:
        break

# Convert number to string for easier digit processing
card_str = str(number)

# Check length and Luhn's Algorithm validity before identifying card type
if valid_length(number):
    if luhns_algorithm(card_str):
        credit_card_type(card_str)
    else:
        print("INVALID")
else:
    print("INVALID")
